app/services/migration/failure_manager.rb

```diff
@@ -0,0 +1,65 @@
+module Migration
+  class FailureManager
+    def record_failure(item, failure_message)
+      return unless data_migration && item.present?
+
+      @item = item
+      @failure_message = failure_message
+
+      write_failure
+      increment_counter
+    end
+
+    def all_failures
+      return unless data_migration
```

Comment:
We could just check `data_migration` is present in the initializer and raise an `ArgumentError` if its `nil` instead of doing it on every instance method